### ISSUE TITLE
FP8 checkpoints store RoPE frequencies quantized — position encoding lands ~2% off (7B) and 5/21 frequencies flush to zero (3B)

### ISSUE THREAD

**Issue by void95** [numz/ComfyUI-SeedVR2_VideoUpscaler#610]
[**### Warning ! Im not a coder i just had an issue using the FP8 Model with weird seams appearing, exactly like here https://www.reddit.com/r/StableDiffusion/comments/1qiq5tg/getting_gridlike_artifactsseams_with_seedvr2/ -> So i decided to fix it  with AI, this was the output AI gave me, and it fixed the code for me. It happened both in comfy and in the extension of forge neo.** 

Hopefully it can help you 

[ISSUE.md](https://github.com/user-attachments/files/30101490/ISSUE.md)

# FP8 checkpoints store RoPE frequencies quantized — position encoding lands ~2% off (7B) and 5/21 frequencies flush to zero (3B)

## Summary

`rope.freqs` is a **persistent** buffer, so it is saved into the checkpoint and carries the
checkpoint's dtype. In the FP8 models that means the RoPE frequencies themselves are stored
as `float8_e4m3fn` — but they are an architecture constant, not a trained weight, and FP8
cannot represent them.

`CompatibleDiT._convert_rope_freqs()` already tries to address this by casting the buffer to
the compute dtype, but that runs *after* the values are loaded. It widens the container and
keeps the error.

Because RoPE is applied per attention window (`self.rope(vid_q, vid_k, window_shape, cache_win)`
in `nablocks/mmsr_block.py`), the distortion repeats with every window, and shows up as steps
on window boundaries. This is very likely the same thing described in the release notes as
"grid-like patterns" on the 7B, which `mixed_block35_fp16` works around.

## The numbers

`rope.py` builds the frequencies from constants only:

```python
# RotaryEmbeddingBase (3B + 7B)
RotaryEmbedding(dim=dim // rope_dim, freqs_for="pixel", max_freq=256)
# MMRotaryEmbeddingBase (3B only)
RotaryEmbedding(dim=dim // rope_dim, freqs_for="lang", theta=10000)
```

With `head_dim=128, rope_dim=3` → `dim=42`, `n=21` frequencies:

| variant | range | stored as FP8 | after the fp8→bf16 cast | stored as FP16 |
|---|---|---|---|---|
| `pixel` (3B + 7B) | 3.14 … 402.12 | **2.295% mean, 4.755% max** | 2.295% mean — *identical* | 0.018% mean |
| `lang` (3B only) | 1.55e-4 … 1.0 | **28.956% mean, 100% max** | 28.956% mean — *identical* | 0.016% mean |

The cast column is not a rounding coincidence: widening an already-quantized value cannot
recover it, so the error after `_convert_rope_freqs()` is bit-for-bit the error before it.

Two concrete failures:

- **`pixel`**: the top frequency is 128·π ≈ 402. `float8_e4m3fn` steps by **32** in that range.
  `302.38` and `282.43` both round to `288.0` — two frequencies collapse into one.

- **`lang`**: the smallest frequency is `1.55e-4`; the smallest positive `float8_e4m3fn`
  subnormal is `1.95e-3`. **5 of 21 frequencies become exactly `0.0`** — those position
  dimensions are dead.

Run [`ISSUE_repro_rope_fp8.py`](#repro) to reproduce. It needs only PyTorch — no GPU, no
model files.

## Observed effect

Reported by a user upscaling 1216×1216 → 2048×2048 with the 7B model. Measured on the output
PNG against a plain bicubic upscale of the same input (so image content cancels out):

```
step at x=960, smooth skin region (rows 1560-1730):
   SeedVR2  : left 107.006 | right 112.285 | step +5.279/255
   bicubic  : left 108.674 | right 109.757 | step +1.083/255
```

`960 = 3 × 320`, i.e. exactly a window boundary. Across all six boundaries the mean step is
**2.03× higher** than at arbitrary columns, and 3 of the 30 strongest columns in the whole
image sit on the 320 grid (chance expectation: 0.10).

**Swapping the FP16 checkpoint in — same image, same seed, same everything — removes the seams
completely.** Rebuilding the frequencies (patch below) removes them with the FP8 checkpoint too.

## Suggested fix

Rebuild the constants instead of widening the quantized values, in
`src/optimization/compatibility.py`:

```python
n = freqs.numel()          # == dim // 2
dim = n * 2
freqs_for = getattr(rope, 'freqs_for', 'pixel')

if freqs_for == 'pixel':
    rebuilt = torch.linspace(1.0, 256 / 2, n, dtype=torch.float32) * math.pi
elif freqs_for == 'lang':
    rebuilt = 1.0 / (10000 ** (torch.arange(0, dim, 2, dtype=torch.float32)[:n] / dim))

rope.freqs.data = rebuilt.to(device=freqs.device)
```

This reproduces the constructor's values exactly (`torch.equal(...) == True` for both
variants). Kept in `float32` rather than the compute dtype on purpose: the angle products
are formed in this buffer's dtype, and `bfloat16` is coarse at 402 (±2 rad).

A cleaner alternative would be `register_buffer("freqs", ..., persistent=False)` in `rope.py`,
so the frequencies are never serialized and are always constructed fresh — but that needs
handling in `meta_non_persistent_buffer_init_fn`, which currently only materializes
`rope.dummy`.

Either way, this would likely make `mixed_block35_fp16` unnecessary as an artifact workaround.

## Environment

- 7B FP8 (`seedvr2_ema_7b_fp8_e4m3fn.safetensors`), also seen on `..._fp16`-mixed variants
- The same artifacts appear in ComfyUI itself, so this is not a port-specific issue
- compute dtype `bfloat16` (confirmed via debug log), RX 9070 XT / ROCm — but the defect is in
  the checkpoint values and is hardware-independent
